sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotNestTypesInArguments.cs

```diff
@@ -67,37 +71,52 @@ node switch
                 _ => Enumerable.Empty<ParameterSyntax>()
             };
 
-        private static int GetNestingDepth(ParameterSyntax parameterSyntax, SyntaxNodeAnalysisContext context) =>
-            GetNestingDepth(context.SemanticModel.GetDeclaredSymbol(parameterSyntax)?.Type, 0 , true);
+        private sealed class GenericWalker : CSharpSyntaxWalker
+        {
+            private static readonly ImmutableArray<KnownType> ignoredTypes =
+                KnownType.SystemFuncVariants
+                         .Union(KnownType.SystemActionVariants)
+                         .Union(new[] { KnownType.System_Linq_Expressions_Expression_T })
+                         .ToImmutableArray();
 
-        private static readonly ImmutableArray<KnownType> expressionFuncActionTypes =
-            KnownType.SystemFuncVariants
-            .Union(KnownType.SystemActionVariants)
-            .Union(new[] { KnownType.System_Linq_Expressions_Expression_T })
-            .ToImmutableArray();
+            private readonly int maxDepth;
+            private readonly SemanticModel model;
 
-        private static int GetNestingDepth(ITypeSymbol argumentSymbol, int depth, bool ignoreExpressions)
-        {
-            var currentOrNestedArgument = argumentSymbol;
-            if (currentOrNestedArgument is IArrayTypeSymbol arrayTypeSymbol)
-            {
-                currentOrNestedArgument = arrayTypeSymbol.ElementType;
-            }
+            private int depth;
+
+            public bool MaxDepthReached { get; private set; }
 
-            if (!(currentOrNestedArgument is INamedTypeSymbol namedTypeSymbol))
+            public GenericWalker(int maxDepth, SemanticModel model)
             {
-                return depth;
+                this.maxDepth = maxDepth;
+                this.model = model;
             }
 
-            var nextDepth = ignoreExpressions && namedTypeSymbol.ConstructedFrom.IsAny(expressionFuncActionTypes) ? depth : depth + 1;
-
-            var typeArguments = namedTypeSymbol.TypeArguments;
-            if (typeArguments != null && typeArguments.Any())
+            public override void VisitGenericName(GenericNameSyntax node)
             {
-                return typeArguments.Max(p => GetNestingDepth(p, nextDepth, false));
-            }
+                if (!(model.GetSymbolInfo(node).Symbol is INamedTypeSymbol namedTypeSymbol))
+                {
+                    return;
+                }
 
-            return depth;
+                if (namedTypeSymbol.ConstructedFrom.IsAny(ignoredTypes))
+                {
+                    base.VisitGenericName(node);
+                }
+                else
+                {
+                    if (depth == maxDepth - 1)
```

Comment:
This condition feels weird. Can we change the initialization constant or do the `- 1` in constructor to have
```C#
if (depth == maxDepth)
{
    MaxDepthReached = true
}
```
Because you reach something once you get there. Current state declares that you're almost there.

Reply:
The depth starts at 0 and increases every time `base.VisitGenericName(node);` is called. In our cases it reaches the max limit since we have to call the method but we exit before doing the incrementation (that's why we have -1 in the condition.). In my opinion starting from -1 would hide this and make the code harder to understand.

Reply:
I didn't want to change the initial value of `depth` but the value of `maxDepth`

Reply:
Like receive `depthCount` in constructor and store `this.maxDepth = depthCount - 1;`